File: src/main/python/AoC2019_08.py
#! /usr/bin/env python3
#
# Advent of Code 2019 Day 8
#

import sys
from typing import cast

from advent_of_code_ocr import convert_6
from aoc.common import InputData
from aoc.common import SolutionBase

WIDTH = 25
HEIGHT = 6

Input = str
Output1 = int
Output2 = str


class Solution(SolutionBase[Input, Output1, Output2]):
    def parse_input(self, input_data: InputData) -> Input:
        return list(input_data)[0]

    def get_layers(self, input: str, width: int, height: int) -> list[str]:
        layer_size = width * height
        return [
            input[i : i + layer_size]  # noqa E203
            for i in range(0, len(input), layer_size)
        ]

    def part_1(self, input: str) -> int:
        layers = self.get_layers(input, WIDTH, HEIGHT)
        least_zeroes = min(layers, key=lambda lyr: lyr.count("0"))
        return least_zeroes.count("1") * least_zeroes.count("2")

    def get_image(self, input: str, width: int, height: int) -> str:
        layers = self.get_layers(input, width, height)
        return "".join(
            next(lyr[i] for lyr in layers if lyr[i] != "2")
            for i in range(0, len(layers[0]))
        )

    def part_2(self, input: str) -> str:
        image = self.get_image(input, WIDTH, HEIGHT)
        to_ocr = "\n".join(
            image[i : i + WIDTH]  # noqa E203
            for i in range(0, WIDTH * HEIGHT, WIDTH)
        )
        return cast(str, convert_6(to_ocr, fill_pixel="1", empty_pixel="0"))

    def samples(self) -> None:
        assert self.get_image("0222112222120000", 2, 2) == "0110"


solution = Solution(2019, 8)


def main() -> None:
    solution.run(sys.argv)


if __name__ == "__main__":
    main()

File: src/main/java/AoC2019_08.java
import static com.github.pareronia.aoc.StringUtils.countMatches;
import static com.github.pareronia.aoc.Utils.toAString;
import static java.util.Comparator.comparing;

import java.util.List;
import java.util.stream.Stream;

import com.github.pareronia.aoc.CharGrid;
import com.github.pareronia.aoc.OCR;
import com.github.pareronia.aoc.solution.SolutionBase;

public class AoC2019_08 extends SolutionBase<String, Integer, String> {
	
	private static final Integer WIDTH = 25;
	private static final Integer HEIGHT = 6;
	
	private AoC2019_08(final boolean debug) {
		super(debug);
	}
	
	public static AoC2019_08 create() {
		return new AoC2019_08(false);
	}

	public static AoC2019_08 createDebug() {
		return new AoC2019_08(true);
	}

	@Override
    protected String parseInput(final List<String> inputs) {
        return inputs.get(0);
    }
	
	private CharGrid parse(
	        final String input, final int width, final int height
	) {
	    return CharGrid.from(input, width * height);
	}

    @Override
	public Integer solvePart1(final String input) {
		final CharGrid layers = parse(input, WIDTH, HEIGHT);
		return layers.getRowsAsStringList().stream()
				.min(comparing(s -> countMatches(s, '0')))
				.map(s -> countMatches(s, '1') * countMatches(s, '2'))
				.orElseThrow();
	}
	
	private String getImage(
	        final String input, final int width, final int height
    ) {
		final CharGrid layers = parse(input, width, height);
		return Stream.iterate(0, i -> i + 1).limit(layers.getWidth())
				.map(i -> layers.getRowsAsStringList().stream()
						.map(row -> row.charAt(i))
						.filter(ch -> ch !=  '2')
						.findFirst()
						.orElseThrow())
				.collect(toAString());
	}

	@Override
	public String solvePart2(final String input) {
		final CharGrid image
		        = CharGrid.from(getImage(input, WIDTH, HEIGHT), WIDTH);
		log(image.replace('1', '\u2592').replace('0', ' '));
		return OCR.convert6(image, '1', '0');
	}

	@Override
    public void samples() {
	    final AoC2019_08 test = AoC2019_08.createDebug();
	    assert test.getImage(TEST, 2, 2).equals("0110");
    }

    public static void main(final String[] args) throws Exception {
        AoC2019_08.create().run();
	}
	
	private static final String TEST = "0222112222120000";
}